### ISSUE TITLE
RubyGems does not activate the default stringio gem on `require "stringio"`

### ISSUE THREAD

**Issue by dazuma** [rubygems/rubygems#7160]
### Describe the problem as clearly as you can

When executing `require "stringio"`, RubyGems does not detect that the (default) `stringio` gem is being loaded. This puts it in an inconsistent state, especially if the `stringio` gem is later loaded explicitly (e.g. via bundler).

### Did you try upgrading RubyGems?

Yes, this is reproducible on Ruby 3.2.2 with RubyGems updated to 3.4.22.

### Post steps to reproduce the problem

Execute the following Ruby script to illustrate the issue:

```ruby
# test.rb
puts "Gems before requiring stringio: #{Gem.loaded_specs.keys.sort}"
require "stringio"
puts "Gems after requiring stringio: #{Gem.loaded_specs.keys.sort}"
```

Result (on an Intel Mac running Sonoma 14.1, but it shouldn't matter):

```
▶ ruby test.rb
Gems before requiring stringio: ["did_you_mean", "error_highlight", "syntax_suggest"]
Gems after requiring stringio: ["did_you_mean", "error_highlight", "syntax_suggest"]
▶
```

I was expecting:

```
Gems before requiring stringio: ["did_you_mean", "error_highlight", "syntax_suggest"]
Gems after requiring stringio: ["did_you_mean", "error_highlight", "stringio", "syntax_suggest"]
```

### Further notes

The ramifications become more visible when you have a newer version of `stringio` installed than the default (3.0.4 on Ruby 3.2.2). As of this writing, the latest is 3.0.9. After `gem install stringio:3.0.9`, then:

```ruby
puts Gem.loaded_specs.keys.sort.inspect  # Does not include stringio
puts require "stringio"                  # prints true
puts Gem.loaded_specs.keys.sort.inspect  # Does not include stringio
puts StringIO::VERSION                   # prints 3.0.4
puts gem "stringio"                      # prints true, suggesting it is now activating the newer gem
puts Gem.loaded_specs.keys.sort.inspect  # Now this does include stringio
puts require "stringio"                  # prints false, because it already thinks it required stringio
puts StringIO::VERSION                   # prints 3.0.4 (not 3.0.9) since it didn't actually load the newer gem
```

This does *not* reproduce for the `set` default gem:

```ruby
puts Gem.loaded_specs.keys.sort.inspect  # Does not include set
puts require "set"                       # prints true
puts Gem.loaded_specs.keys.sort.inspect  # DOES include set
puts gem "set"                           # prints false, since the gem is already activated
```

I was showing this to Samuel briefly at RubyConf. His opinion is that, since the actual file loaded when requiring "stringio" is a built shared library (e.g. `stringio.bundle` or `stringio.so`), it is not covered in the gemspec, hence the `Kernel#require` additions fail to detect that it is actually part of a gem, and hence fail to activate the gem.

### Run `gem env` and paste the output below

```
▶ gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 3.4.22
  - RUBY VERSION: 3.2.2 (2023-03-30 patchlevel 53) [x86_64-darwin22]
  - INSTALLATION DIRECTORY: /Users/dazuma/opt/asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0
  - USER INSTALLATION DIRECTORY: /Users/dazuma/.gem/ruby/3.2.0
  - RUBY EXECUTABLE: /Users/dazuma/opt/asdf/installs/ruby/3.2.2/bin/ruby
  - GIT EXECUTABLE: /usr/local/bin/git
  - EXECUTABLE DIRECTORY: /Users/dazuma/opt/asdf/installs/ruby/3.2.2/bin
  - SPEC CACHE DIRECTORY: /Users/dazuma/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /Users/dazuma/opt/asdf/installs/ruby/3.2.2/etc
  - RUBYGEMS PLATFORMS:
     - ruby
     - x86_64-darwin-22
  - GEM PATHS:
     - /Users/dazuma/opt/asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0
     - /Users/dazuma/.gem/ruby/3.2.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
     - :benchmark => false
     - :sources => ["http://rubygems.org/"]
     - "gemcutter_key" => "********************************"
     - "gem" => "--no-ri --no-rdoc"
     - "install" => "--no-document"
     - "update" => "--no-document"
  - REMOTE SOURCES:
     - http://rubygems.org/
  - SHELL PATH:
     - /Users/dazuma/opt/asdf/installs/ruby/3.2.2/bin
     - /Users/dazuma/opt/google-cloud-sdk/bin
     - /Users/dazuma/opt/asdf/shims
     - /Users/dazuma/opt/asdf/bin
     - /Users/dazuma/bin
     - /usr/local/bin
     - /usr/local/sbin
     - /usr/bin
     - /bin
     - /usr/sbin
     - /sbin
▶
```

**Comment by deivid-rodriguez** [comment]
Sorry we never got back here! This should get fixed by https://github.com/ruby/ruby/pull/9673. As a temporary workaround, you could manually add `stringio.so` to the list of files in `/Users/dazuma/opt/asdf/installs/ruby/3.2.2/lib/ruby/gems/3.2.0/specifications/default/stringio-3.0.4.gemspec` and it should start working as expected.

**Comment by deivid-rodriguez** [comment]
This should be fixed in Ruby 3.4.0-dev now, so I'll close this. Let us know if it's still not working though!
